### PR TITLE
Rotation Capability

### PR DESCRIPTION
**Pull request by chloeqjz**
### Description
added rotation capability functions to allow a user to offer their rotation capability to another account, and allow another account to accept a rotation capability.

### Test Plan
unit tests

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/3091)

### CODE REVIEW

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 1]
```diff
@@ -46,7 +46,7 @@ module aptos_framework::account {
     }
 
     struct CapabilityOffer<phantom T> has store { for: Option<address> }
-    struct RotationCapability has drop, store { account: address }
+    struct RotationCapability has key, drop, store { account: address }
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:49]: I didn't want to make this key. I intentionally want folks to think about where they want to place this and not make everything a resource. We can go into details tomorrow.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 2]
```diff
@@ -63,6 +63,11 @@ module aptos_framework::account {
         new_public_key: vector<u8>,
     }
 
+    struct RotationCapabilityOfferProof has drop {
+        sequence_number: u64,
+        intended_receiver_address: address,
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:68]: maybe just recipient_address

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 3]
```diff
@@ -102,6 +108,14 @@ module aptos_framework::account {
     const PROLOGUE_ESEQUENCE_NUMBER_TOO_BIG: u64 = 1011;
     const PROLOGUE_ESECONDARY_KEYS_ADDRESSES_COUNT_MISMATCH: u64 = 1012;
 
+    #[test_only]
+    const OFFER_ROTATION_CAPABILITY_SIGNATURE: vector<u8> = vector<u8>[128, 67, 218, 117, 250, 136, 201, 53, 214, 46, 129, 120, 102, 249, 108, 116, 158, 237, 0, 162, 227, 205, 107, 230, 159, 155, 102, 196, 166, 224, 135, 72, 138, 200, 37, 98, 140, 94, 109, 164, 48, 26, 239, 162, 140, 181, 202, 255, 87, 231, 63, 198, 50, 205, 207, 206, 89, 62, 58, 190, 116, 92, 179, 12];
+    #[test_only]
+    const OFFER_ROTATION_CAPABILITY_INVALID_SIGNATURE: vector<u8> = vector<u8>[100, 67, 218, 117, 250, 136, 201, 53, 214, 46, 129, 120, 102, 249, 108, 116, 158, 237, 0, 162, 227, 205, 107, 230, 159, 155, 102, 196, 166, 224, 135, 72, 138, 200, 37, 98, 140, 94, 109, 164, 48, 26, 239, 162, 140, 181, 202, 255, 87, 231, 63, 198, 50, 205, 207, 206, 89, 62, 58, 190, 116, 92, 179, 12];
+
+    #[test_only]
+    const OFFER_ROTATION_CAPABILITY_PUBLIC_KEY: vector<u8> = vector<u8>[246, 107, 240, 206, 92, 235, 88, 43, 147, 214, 120, 8, 32, 194, 2, 91, 153, 103, 174, 218, 162, 89, 189, 187, 159, 61, 2, 151, 236, 237, 14, 24];
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:117]: you can use something like x"hexdigits", see https://move-language.github.io/move/vector.html?highlight=hex%20string#vectoru8-literals

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 4]
```diff
@@ -256,6 +270,43 @@ module aptos_framework::account {
...
+            error::invalid_argument(EMALFORMED_PROOF_OF_KNOWLEDGE)
+        );
+
+        let account_resource = borrow_global_mut<Account>(addr);
+
+        let rotation_capability_offer_proof = RotationCapabilityOfferProof {
+            sequence_number: account_resource.sequence_number,
+            intended_receiver_address,
+        };
+
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:291]: you need to verify that the public key is the same as the auth key
also note that this only works with the ed25519 key, so we probably want to call this offer_rotation_capability_ed25519

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 5]
```diff
@@ -256,6 +270,43 @@ module aptos_framework::account {
...
+        let addr = signer::address_of(account);
+        assert!(exists_at(addr) && exists_at(offerer_address), error::not_found(EACCOUNT));
+
+        let account_resource = borrow_global_mut<Account>(offerer_address);
+        assert!(option::contains(&account_resource.rotation_capability_offer.for, &addr), EINVALID_ACCEPT_ROTATION_CAPABILITY);
+
+        let rotation_capability = RotationCapability {
+            account: offerer_address,
+        };
+        move_to(account, rotation_capability);
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:306]: just return it, no need to store it, that is something we can do in the test code

**Review by davidiw — COMMENTED** [review]
looking pretty good :)

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 6]
```diff
@@ -272,6 +286,57 @@ module aptos_framework::account {
...
+        // Get the current authentication key of the account
+        let account_resource = borrow_global_mut<Account>(addr);
+        let curr_auth_key = create_address(account_resource.authentication_key);
+
+        // Check curr_auth_key matches expected_current_auth_key:
+        //  1. First, append the Ed25519 scheme identifier '0x00' to the PK
+        //  2. Second, hash this using SHA3-256
+        vector::push_back(&mut pk_bytes, 0);
+        let expected_current_auth_key = hash::sha3_256(pk_bytes);
+        assert!(create_address(expected_current_auth_key) == curr_auth_key, std::error::unauthenticated(EWRONG_CURRENT_PUBLIC_KEY));
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:306]: We're using this twice in this module. Can we define a function? Can it be public so other modules can use it?

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 7]
```diff
@@ -272,6 +286,57 @@ module aptos_framework::account {
         account_resource.authentication_key = new_auth_key;
     }
 
+    // Offer rotation capability of this account to another address
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:289]: Use three slash /// comments for top-level function comments. I would write account as `account`, to more clearly indicate the variable representing the account.

Comment (alinush) [reply aptos-move/framework/aptos-framework/sources/account.move:289]: Also, explain a little what the other three args are.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 8]
```diff
@@ -272,6 +286,57 @@ module aptos_framework::account {
...
+    public entry fun offer_rotation_capability_ed25519(account: &signer, rotation_capability_sig_bytes: vector<u8>, pk_bytes: vector<u8>, recipient_address: address) acquires Account {
```
Comment (alinush) [inline aptos-move/framework/aptos-framework/sources/account.move:290]: Rename `pk_bytes` to `account_pk_bytes` to better indicate that this is the account's PK.

**Review by alinush — COMMENTED** [review]
Not sure I understand what the feature is here. Is this just a draft PR or is it finalized?

Specifically, `offer_rotation_capability_ed25519` fills in the `rotation_capability_offer` field of the **offerer account** with the **recipient's** address, assuming the appropriate signatures verify.

Then, `accept_rotation_capability_ed25519` is called by the _recipient_ account to check if the offerer is delegating the rotation of their account to the recipient: i.e., to check if the offerer's account resource has its `rotation_capability_offer.for` field set to the recipient's address.

If so, it returns a `RotationCapability` struct with the address of the offerer, which I suppose can be later used by the recipient to rotate the account.

However, I can't find any logic using this capability...

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 6 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:306]: yea good point, will add a function for this

File: aptos-move/framework/aptos-framework/sources/account.move (hunk 7 above)
Comment (chloeqjz) [reply aptos-move/framework/aptos-framework/sources/account.move:289]: ok makes sense - will add more comments

**Comment by chloeqjz** [comment]
> Not sure I understand what the feature is here. Is this just a draft PR or is it finalized?
> 
> Specifically, `offer_rotation_capability_ed25519` fills in the `rotation_capability_offer` field of the **offerer account** with the **recipient's** address, assuming the appropriate signatures verify.
> 
> Then, `accept_rotation_capability_ed25519` is called by the _recipient_ account to check if the offerer is delegating the rotation of their account to the recipient: i.e., to check if the offerer's account resource has its `rotation_capability_offer.for` field set to the recipient's address.
> 
> If so, it returns a `RotationCapability` struct with the address of the offerer, which I suppose can be later used by the recipient to rotate the key of the offerer's account.
> 
> However, I can't find any logic using this capability...

Yea the other things can happen off-chain? or we can add event to notify users later

**Comment by alinush** [comment]
I think the actual key rotation using this capability has to happen on
chain via some Move code, which I don’t see in this PR.

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 9]
```diff
@@ -102,6 +109,13 @@ module aptos_framework::account {
     const PROLOGUE_ESEQUENCE_NUMBER_TOO_BIG: u64 = 1009;
     const PROLOGUE_ESECONDARY_KEYS_ADDRESSES_COUNT_MISMATCH: u64 = 1010;
 
+    #[test_only]
+    const OFFER_ROTATION_CAPABILITY_SIGNATURE: vector<u8> = x"68f7d09ef7a9d8d7450d600b10231e6512610f919a63bd71bea1c907f7e101ed333bff360eeda97a8637a53fd622d597c03a0d6fd1315c6fa23719983ff7de0c";
+    #[test_only]
+    const OFFER_ROTATION_CAPABILITY_INVALID_SIGNATURE: vector<u8> = x"78f7d09ef7a9d8d7450d600b10231e6512610f919a63bd71bea1c907f7e101ed333bff360eeda97a8637a53fd622d597c03a0d6fd1315c6fa23719983ff7de0c";
+    #[test_only]
+    const OFFER_ROTATION_CAPABILITY_PUBLIC_KEY: vector<u8> = x"f66bf0ce5ceb582b93d6780820c2025b9967aedaa259bdbb9f3d0297eced0e18";
+
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:118]: maybe move below to the tests themselves.

`verify_authentication_key_matches_ed25519_public_key`

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 10]
```diff
@@ -209,6 +223,15 @@ module aptos_framework::account {
         account_resource.authentication_key = new_auth_key;
     }
 
+    // Check account_pk_bytes matches current_auth_key:
+    //  1. First, append the Ed25519 scheme identifier '0x00' to `account_pk_bytes`
+    //  2. Second, hash this using SHA3-256
+    public fun verify_authentication_key_matches_public_key(account_auth_key: vector<u8>, account_pk_bytes: vector<u8>) : bool {
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:229]: let's not make this public and we should specialize it as ed25519

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 11]
```diff
@@ -209,6 +223,15 @@ module aptos_framework::account {
...
+        vector::push_back(&mut account_pk_bytes, 0);
+        let expected_account_auth_key = hash::sha3_256(account_pk_bytes);
+        create_address(expected_account_auth_key) == create_address(account_auth_key)
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:232]: you don't need to call create_address to make the comparison. just compare them directly

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 12]
```diff
@@ -229,22 +252,16 @@ module aptos_framework::account {
...
-        assert!(create_address(expected_current_auth_key) == curr_auth_key, std::error::unauthenticated(EWRONG_CURRENT_PUBLIC_KEY));
+        assert!(verify_authentication_key_matches_public_key(account_resource.authentication_key, curr_pk_bytes), std::error::unauthenticated(EWRONG_CURRENT_PUBLIC_KEY));
 
+        let curr_auth_key = create_address(account_resource.authentication_key);
         // Construct a RotationProofChallenge struct
         let challenge = RotationProofChallenge {
             sequence_number: account_resource.sequence_number,
             originator: addr,
-            current_auth_key: curr_auth_key,
+            current_auth_key:  curr_auth_key,
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:264]: nit unnecessary whitespace

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 13]
```diff
@@ -272,6 +289,53 @@ module aptos_framework::account {
         account_resource.authentication_key = new_auth_key;
     }
 
+    /// Offer rotation capability of this account to another address
+    /// To authorize the rotation capability offer, a signature under the current public key on a `RotationCapabilityOfferProofChallenge`
+    /// is given in `rotation_capability_sig_bytes`. The current public key is passed into `account_pk_bytes` to verify proof-of-knowledge.
+    /// The recipient address refers to the address that the account owner wants to give the rotation capability to.
+    public entry fun offer_rotation_capability_ed25519(account: &signer, rotation_capability_sig_bytes: vector<u8>, account_pk_bytes: vector<u8>, recipient_address: address) acquires Account {
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:296]: pk -> public_key
pk -> private_key 😅

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 14]
```diff
@@ -272,6 +289,53 @@ module aptos_framework::account {
...
+        assert!(ed25519::signature_verify_strict_t(&rotation_capability_sig, &pubkey, rotation_capability_offer_proof_challenge), EINVALID_PROOF_OF_KNOWLEDGE);
+
+        // Add the recipient's address in account owner's rotation capability offer once we verify that this action is intended by the account owner
+        option::fill(&mut account_resource.rotation_capability_offer.for, recipient_address);
+    }
+
+    // Accept rotation capability from `offerer_address` if there's an existing rotation capability offer to the account owner in the offerer's account
+    public fun accept_rotation_capability_ed25519(account: &signer, offerer_address: address) : RotationCapability acquires Account {
+        let addr = signer::address_of(account);
+        assert!(exists_at(addr) && exists_at(offerer_address), error::not_found(EACCOUNT_DOES_NOT_EXIST));
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:323]: you should only check the offerer not the recipient

File: aptos-move/framework/aptos-framework/sources/account.move [hunk 15]
```diff
@@ -605,4 +669,62 @@ module aptos_framework::account {
         let pk = x"0000000000000000000000000000000000000000000000000000000000000000";
         rotate_authentication_key_ed25519(&alice, test_signature, test_signature, pk, pk);
     }
+
+    #[test(bob = @0x345)]
+    public entry fun test_valid_offer_rotation_capability(bob: signer) acquires Account {
```
Comment (davidiw) [inline aptos-move/framework/aptos-framework/sources/account.move:674]: this test is a smaller version of `test_valid_accept_rotation_capability`, let's delete it.

**Review by davidiw — APPROVED** [review]
looks great!
